### ISSUE TITLE
Bootstrapper

### ISSUE THREAD

**Issue by SiyataDishmaya** [oleg-shilo/wixsharp#393]
I based my installer on the code of the simplified bootstapper. The first install calls a second msi (which was also built with WixSharp). When running the install, I get the following error. 

"Another installation is in progress. You must complete that installation before continuing this one."

There is a limitation that only one msi can run at any one moment. Did the sample ever run? How do I get around this?
Thanks

**Comment by oleg-shilo** [comment]
Yes the sample was run after is was developed. I have also also tried it before responding to your post. All works as expected. The sample installs both "MyProduct" and "FakeCRT" products.

![image](https://user-images.githubusercontent.com/16729806/41413189-a846daf2-7025-11e8-8379-e0c281781276.png)

The only bit that is not accurate in the sample is the implementation of `IsCRTInstalled`, which always returns false. But it is irrelevant for the objective of the sample.

You are experiencing problem with some MSI installation being in the active phase while your test is trying to start another one. My speculation that it is some other abandoned install in progress. 

I suggest you kill all `msiexec.exe` instances and repeat the test. I just don't see how the sample can proceed to the second msi without waiting for the first one (CRTMsiFile) to finish:
```c#
Process.Start(CRTMsiFile)
       .WaitForExit();
```

Anyway, test it again and the chances are that it will work. 

BTW there is another bootstrapping technique is coming with the very next release (~next week). See https://github.com/oleg-shilo/wixsharp/tree/master/Source/src/WixSharp.Samples/Wix%23%20Samples/Bootstrapper/NsisBootstrapper. Thanks @TigranAnchorfree WixSharp can now use NSIS bootstrapping engine.

**Comment by Xaddan** [comment]
> The first install calls a second msi..

The second installer will not work until the first one finishes its work.
It is necessary to call both installers from the bootstrapper.

**Comment by oleg-shilo** [comment]
I think there is a confusion here. 

With MSI session has two distinctive phases. First one is UI phase of interactioon with the user. This phase is not atomic and one can start as many msi setups and as amny products he wants. They all will start and wait for the user input. 

The second phase is the actual installation and it is atomic - MSI runtime allows onl one to be active at the time. 

Simplified bootstrapper uses the first phase to analyze the system and start the "second" setup. Then it waits for that second setup to finish. And only after that proceeds to its own installation. Thus there is no any collision between the "second" installation phases of these two setups.

This is what my test just confirmed.

Now, did you have the opportunity to repeat the test as I suggested?

**Comment by SiyataDishmaya** [comment]
I used your sample and added my second install as the Fake CRT.msi and it worked! The two critcal conditions for it to succed is that you need to set it up for Step.LaunchConditions and Sequence.InstallUISequence. If not, the install will not work.. 
Thank you for your input.

**Comment by shengxue** [comment]
@SiyataDishmaya @oleg-shilo could you please share your code here? Thanks!

**Comment by oleg-shilo** [comment]
https://github.com/oleg-shilo/wixsharp/tree/d809ad85510a84849ee0ab0a249d375077c64ef5/Source/src/WixSharp.Samples/Wix%23%20Samples/Bootstrapper/Simplified%20Bootstrapper

**Comment by oleg-shilo** [comment]
You can

1. Simplified Bootstrapper: instead of `@"Fake CRT.msi"` use `"vc_redist.x64.exe"`
    Follow the code and update all the references in the [setup.cs](https://github.com/oleg-shilo/wixsharp/blob/d809ad85510a84849ee0ab0a249d375077c64ef5/Source/src/WixSharp.Samples/Wix%23%20Samples/Bootstrapper/Simplified%20Bootstrapper/setup.cs) file. 
    Specifically, line 24 and 45.

2. `ProjectOnAfterInstall` approach works just fine. you only need to ensure that your vc_redist.x64.exe starts only after your primary msi exits. It is a matter of a simple inter-process synchronization.

**Comment by shengxue** [comment]
@oleg-shilo The below code show that my primary msi always exits.  Maybe this is the reason why I got error "Another installation is in progress.... ".    I didn't get your idea of ```you only need to ensure that your vc_redist.x64.exe starts only after your primary msi exits```.  

It looks to me that I should instead ensure that my vc_redist.x64.exe starts only after the installation of my primary msi finishes.


``` csharp
            project.AfterInstall += e =>
            {
                    System.IO.File.AppendAllText(@"c:\debug.txt", "Install 3rd party softwares" + Environment.NewLine);
                    int i = 0;
                    while (System.IO.File.Exists(e.MsiFile) && i < 10)
                    {
                        Thread.Sleep(10);
                        System.IO.File.AppendAllText(@"c:\debug.txt", $"e.MsiFile exists" + Environment.NewLine);
                        i++;
                    }
            
                    InstallVcRedist();
            };
```

**Comment by shengxue** [comment]
@oleg-shilo  Actually, I tried ```project.UIInitialized```, and it works well at non-silent mode.  
Unfortunately, when the msi runs in command window with option ```/quiet```,  ```project.UIInitialized``` event handler is not called.

``` csharp
            project.UIInitialized += (SetupEventArgs e) =>
            {
                   InstallVcRedist();
                   InstallDotNet5();
            }
```

**Comment by shengxue** [comment]
@oleg-shilo  are there some other event handler similar to ```UIInitialized``` that happens before the installation of msi? and also triggered in silent mode?

**Comment by oleg-shilo** [comment]
No there is none. From the moment MSI fires any event, it is already loaded and the installation is in progress.

>starts only after the installation of my primary msi finishes.

I am afraid you will not be able to finish as you are waiting in the AfterInstall which will always return before your install is finished. Still.... makes sense to check

What you can do instead is start from AfterInstall a simple launcher all that will wait for your primary install to finish and then will start the installation of the crt.

This launcher can be the same assambly that implements your managed setup (pseudo-code):
```C#
    static void Main(string[] args )
    {
        if (args.Contain("-launch"))
        {
             var processToWait = args[2]; 
             var msiToStart = args[1]; 
             while (Process.GetProcessById(processToWait) != null)
             {
                 Thread.Sleep (1000);
             }
             Process.Strart("msiexec", msiToStart);
             return;
        }
    
    . . .
    project.AfterInstall += e =>
    {
          string launcher = System.Reflectioon.Assembly.ExecutingAssembly.Location;
          string crtMsi = ExtractVcRedistTo("<temp dir>");
          int primarySetupProcess = Process.CurrrentProcess.Id;
          
         Process.Start(launcher, $"-launch \"{crtMsi}\" {primarySetupProcess }");
    };

**Comment by shengxue** [comment]
@oleg-shilo  Regarding the ```SimplifiedBoostrapper-CRT```,  if I use ```ManagedProject``` instead, assign ```ManageUI``` to the ```ManagedProject```, the  ```Sequence.InstallUISequence``` action never get triggered.  But if I don't assign ```ManageUI``` to the ```ManagedProject```, the  ```Sequence.InstallUISequence``` action is triggered.

any idea to make the action triggered even when ```ManageUI``` is defined?

**Comment by oleg-shilo** [comment]
The recommended way is not to use custom actions as I illustrated in the sample above. This way you can rely on the power of Managed UI and managed events, which are by far superior to any MSI/WiX custom actions support.

BTW, I just published a workaround that allow showing ManagedUI from the WiX Bootstrapper. I think you might be interested in that.

https://github.com/oleg-shilo/wixsharp/tree/master/Source/src/WixSharp.Samples/Wix%23%20Samples/Bootstrapper/WixBootstrapper_EmbeddedUI
